iceoryx_posh/test/moduletests/test_popo_publisher_port.cpp

```diff
@@ -0,0 +1,390 @@
+    iox::popo::PublisherPortUser m_publisherPortUser{&m_publisherPortData};
+};
+
+TEST_F(PublisherPort_test, initial_state)
+{
+    EXPECT_FALSE(m_publisherPortUser.isOffered());
+    EXPECT_FALSE(m_publisherPortUser.hasSubscribers());
+    auto maybeCaproMessage = m_publisherPortRouDi.getCaProMessage();
+    EXPECT_FALSE(maybeCaproMessage.has_value());
+}
+
+TEST_F(PublisherPort_test, offering_stop_offering)
+{
+    // initial state
+    EXPECT_FALSE(m_publisherPortUser.isOffered());
+    auto maybeCaproMessage = m_publisherPortRouDi.getCaProMessage();
+    EXPECT_FALSE(maybeCaproMessage.has_value());
+
+    // offering
+    m_publisherPortUser.offer();
+    EXPECT_TRUE(m_publisherPortUser.isOffered());
+    maybeCaproMessage = m_publisherPortRouDi.getCaProMessage();
+    EXPECT_TRUE(maybeCaproMessage.has_value());
+    auto caproMessage = maybeCaproMessage.value();
+    EXPECT_THAT(caproMessage.m_type, Eq(iox::capro::CaproMessageType::OFFER));
+    EXPECT_THAT(caproMessage.m_serviceDescription, Eq(iox::capro::ServiceDescription("a", "b", "c")));
+    EXPECT_THAT(caproMessage.m_subType, Eq(iox::capro::CaproMessageSubType::EVENT));
+    EXPECT_THAT(caproMessage.m_historyCapacity, Eq(0u));
+
+    // stop offering
+    m_publisherPortUser.stopOffer();
+    EXPECT_FALSE(m_publisherPortUser.isOffered());
+    maybeCaproMessage = m_publisherPortRouDi.getCaProMessage();
+    EXPECT_TRUE(maybeCaproMessage.has_value());
+    caproMessage = maybeCaproMessage.value();
+    EXPECT_THAT(caproMessage.m_type, Eq(iox::capro::CaproMessageType::STOP_OFFER));
+    EXPECT_THAT(caproMessage.m_serviceDescription, Eq(iox::capro::ServiceDescription("a", "b", "c")));
+}
+
+TEST_F(PublisherPort_test, offering_with_history)
+{
+    iox::popo::PublisherPortData m_publisherPortDataHistory{iox::capro::ServiceDescription("x", "y", "z"),
+                                                            "myApp",
+                                                            &m_memoryManager,
+                                                            iox::MAX_HISTORY_CAPACITY_OF_CHUNK_DISTRIBUTOR};
+    iox::popo::PublisherPortUser m_publisherPortUserHistory{&m_publisherPortDataHistory};
+    iox::popo::PublisherPortRouDi m_publisherPortRouDiHistory{&m_publisherPortDataHistory};
+
+    // offering
+    m_publisherPortUserHistory.offer();
+    EXPECT_TRUE(m_publisherPortUserHistory.isOffered());
+    auto maybeCaproMessage = m_publisherPortRouDiHistory.getCaProMessage();
+    EXPECT_TRUE(maybeCaproMessage.has_value());
+    auto caproMessage = maybeCaproMessage.value();
+    EXPECT_THAT(caproMessage.m_type, Eq(iox::capro::CaproMessageType::OFFER));
+    EXPECT_THAT(caproMessage.m_serviceDescription, Eq(iox::capro::ServiceDescription("x", "y", "z")));
+    EXPECT_THAT(caproMessage.m_subType, Eq(iox::capro::CaproMessageSubType::FIELD));
+    EXPECT_THAT(caproMessage.m_historyCapacity, Eq(iox::MAX_HISTORY_CAPACITY_OF_CHUNK_DISTRIBUTOR));
+}
+
+TEST_F(PublisherPort_test, allocate_free)
```

Comment:
These are again multiple tests:
- no used chunks on startup
- allocate available chunks returns valid chunkHeader
- allocating chunk reduces amount of available chunks
- freeing chunk increases amount of available chunks

I think here are also some test cases missing. Think of the rule: Zero, One, Many, CornerCases
Zero would be initial state
One is tested here
Many, can we allocate multiple chunks?
CornerCases, does the maximum allowed number in our system work. How many chunks is a publisher allowed to allocate -> this has to work. Another test would be, allocate one more chunk then allowed, this has to fail

Reply:
Valid point, But some of them are already addressed on the lower layer. A lo calls are just forwarded and I concentrated on things that really happen in this class. But I will check again